File: ruby/lib/cucumber/ci_environment.rb
require 'uri'
require 'json'
require 'cucumber/ci_environment/variable_expression'

module Cucumber
  module CiEnvironment
    extend VariableExpression
    CI_ENVIRONMENTS_PATH = File.join(File.dirname(__FILE__), 'ci_environment/CiEnvironments.json')

    def detect_ci_environment(env, file_reader = IO.method(:read))
      ci_environments = JSON.parse(file_reader.call(CI_ENVIRONMENTS_PATH))
      ci_environments.each do |ci_environment|
        detected = detect(ci_environment, env, file_reader)
        return detected unless detected.nil?
      end

      nil
    end

    def detect(ci_environment, env, file_reader)
      url = evaluate(ci_environment['url'], env)
      return nil if url.nil?

      result = {
        name: ci_environment['name'],
        url: url,
        buildNumber: evaluate(ci_environment['buildNumber'], env),
      }

      detected_git = detect_git(ci_environment, env, file_reader)
      result[:git] = detected_git if detected_git
      result
    end

    def detect_git(ci_environment, env, file_reader)
      revision = detect_revision(ci_environment, env, file_reader)
      return nil if revision.nil?

      remote = evaluate(ci_environment['git']['remote'], env)
      return nil if remote.nil?

      git_info = {
        remote: remove_userinfo_from_url(remote),
        revision: revision,
      }

      tag = evaluate(ci_environment['git']['tag'], env)
      branch = evaluate(ci_environment['git']['branch'], env)
      git_info[:tag] = tag if tag
      git_info[:branch] = branch if branch
      git_info
    end

    def detect_revision(ci_environment, env, file_reader)
      if env['GITHUB_EVENT_NAME'] == 'pull_request'
        raise StandardError('GITHUB_EVENT_PATH not set') unless env['GITHUB_EVENT_PATH']
        event = JSON.parse(file_reader.call(env['GITHUB_EVENT_PATH']))
        raise StandardError("No after property in #{env['GITHUB_EVENT_PATH']}:\n#{JSON.pretty_generate(event)}") unless event['after']
        return event['after']
      end

      return evaluate(ci_environment['git']['revision'], env)
    end

    def remove_userinfo_from_url(value)
      return nil if value.nil?

      begin
        uri = URI(value)
        uri.userinfo = ''
        uri.to_s
      rescue StandardError
        value
      end
    end

    module_function :detect_ci_environment, :detect, :detect_git, :detect_revision, :remove_userinfo_from_url
  end
end

File: python/src/ci_environment/ci_environment.py
from collections import namedtuple
from contextlib import suppress
from functools import partial, wraps
from itertools import filterfalse
from json import loads as loads_json
from operator import is_
from pathlib import Path
from pprint import pformat
from typing import Dict
from urllib.parse import urlparse

from importlib_resources import as_file, files

from ci_environment.variable_expression import evaluate

_key_value = namedtuple("_key_value", ["key", "value"])


@wraps(dict)
def _opt_dict(*args, **kwargs) -> Dict:
    return {k: v for k, v in dict(*args, **kwargs).items() if v is not None}


def _flip(func):
    def wrapped(*args, **kwargs):
        if len(args) > 1:
            first, *other, last = args
            return func(last, *other, first, **kwargs)
        else:
            return func(*args, **kwargs)

    return wrapped


def detect_ci_environment(env: Dict[str, str]):
    with as_file(files("ci_environment").joinpath("CiEnvironments.json")) as path:
        ci_environments = loads_json(Path(path).read_text())

    return next(filterfalse(partial(_flip(is_), None), map(partial(detect, env=env), ci_environments)), None)


def detect(ci_environment, env: Dict[str, str]):
    if (url := evaluate(ci_environment["url"], env)) is not None:
        return dict(
            name=ci_environment["name"],
            url=url,
            buildNumber=evaluate(ci_environment["buildNumber"], env),
            **_opt_dict(git=detect_git(ci_environment, env)),
        )


def detect_git(ci_environment, env: Dict[str, str]):
    if all(
        [
            (revision := detect_revision(ci_environment, env)) is not None,
            (remote := evaluate(ci_environment["git"]["remote"], env)) is not None,
        ]
    ):
        return dict(
            remote=remove_userinfo_from_url(remote),
            revision=revision,
            **_opt_dict(
                tag=evaluate(ci_environment.get("git", {}).get("tag"), env),
                branch=evaluate(ci_environment.get("git", {}).get("branch"), env),
            ),
        )


def detect_revision(ci_environment, env: Dict[str, str]):
    if env.get("GITHUB_EVENT_NAME") == "pull_request":
        try:
            event = loads_json(Path(env["GITHUB_EVENT_PATH"]).read_text())
        except KeyError as e:
            raise ValueError from e
        try:
            revision = event["pull_request"]["head"]["sha"]
        except KeyError as e:
            raise ValueError(pformat(event)) from e
        else:
            return revision
    else:
        return evaluate(ci_environment["git"]["revision"], env)


def remove_userinfo_from_url(remote):
    if remote is not None:
        with suppress(Exception):
            return (parsed_url := urlparse(remote))._replace(netloc=parsed_url.hostname).geturl()
        return remote
